src/www/ui_components/server_connection_viz.ts

```diff
@@ -242,62 +231,61 @@ Polymer({
       <img id="large-zero" src\$="[[rootPath]]assets/disc_empty.png" class\$="{{state}}">
       <img id="large" src\$="[[rootPath]]assets/disc_color.png" class\$="green {{animationState}}">
     </div>
-`,
+  `;
 
-  is: 'server-connection-viz',
+  animationStartMS: number;
 
-  properties: {
-    rootPath: String,
-    state: {
-      type: String,
-      observer: '_onStateChanged',
-    },
-    animationState: {
-      type: String,
-      notify: true,
-      value: 'ZERO_STATE',
-    },
-    expanded: {
-      type: Boolean,
-      value: false,
-    },
-  },
+  @property({type: String}) rootPath: string;
+  @property({type: Boolean}) expanded: boolean;
 
-  _onStateChanged: function _onStateChanged(newState) {
-    this._updateAnimationState();
-  },
+  @property({type: String, observer: 'syncAnimationState'}) state: ServerConnectionState;
+  @property({type: String}) animationState: ServerConnectionState = ServerConnectionState.INITIAL;
 
-  // Update CSS when modifying the animation duration.
-  _ANIMATION_DURATION_MS: 1750,
+  @computed('expanded')
+  get expandedClassName() {
+    return this.expanded ? 'expanded' : '';
+  }
 
-  _animationStartMs: null,
+  @computed('state')
+  get shouldAnimate() {
+    return this.isAnimationState(this.state);
+  }
+
+  @computed('animationState')
+  get isAnimating() {
+    return this.isAnimationState(this.animationState);
+  }
 
-  _updateAnimationState: function() {
-    if (this._isAnimating(this.state)) {
-      this._animationStartMs = new Date().getTime();
+  protected syncAnimationState() {
+    if (this.shouldAnimate) {
+      return this.startAnimation();
     }
-    if (this.state === 'CONNECTED' || this.state === 'DISCONNECTED') {
-      if (this._isAnimating(this.animationState)) {
-        var now = new Date().getTime();
-        var elapsedAnimationMs = now - this._animationStartMs;
-        var remainingAnimationMs =
-            this._ANIMATION_DURATION_MS - (elapsedAnimationMs % this._ANIMATION_DURATION_MS);
-        // Update the state only after the animation cycle has finished to avoid jumpiness.
-        var _this = this;
-        this.async(function() {
-          _this.animationState = _this.state;
-        }, remainingAnimationMs);
-        return;
-      }
+
+    if (!this.shouldAnimate && this.isAnimating) {
+      return this.stopAnimation();
     }
+
+    this.animationState = this.state;
+  }
+
+  private startAnimation() {
+    this.animationStartMS = Date.now();
+
     this.animationState = this.state;
```

Comment:
I have the impression that this component doesn't need to know about `state` it only needs animation on and animation off, right? Perhaps a single boolean property and push the state logic to the card?

Reply:
Yeah, I like that; let me try it

Reply:
Ah, actually, it animates differently based on state - if it's `DISCONNECTING` the animation plays in reverse:

```css
      #large.CONNECTING,
      #large-grey.CONNECTING,
      #large.RECONNECTING,
      #large-grey.RECONNECTING {
        animation: rotate-with-pause 1.75s ease-out 500ms infinite;
      }
      #small.DISCONNECTING,
      #small-grey.DISCONNECTING {
        animation: rotate-backward-with-pause 1.75s ease-out infinite;
      }
```